refactor!: Socket.IO

gavindsouza opened the pull request:
### Rationale

Giving data to only those who need it, and can access it: many smaller size rooms, lesser noise.

### Main Changes

- Each user gets added to their own "user" & the common "website" room by default 
- Maintain separate rooms for each Document's `list_update` and `doc_update` events
- Opening list or report views subscribes to the list_update event for said DocType, if the user is allowed to view it
- Docinfo and doc_update events (same room) are relayed only if user has access to room / Document
- Rename event `f"update_docinfo_for_{self.reference_doctype}_{self.reference_name}"` to `"docinfo_update"`
- Allow only System Users access to the site room (as opposed to all users) 
- Split authentication and socket request validations into middleware

### Other Changes

- Removed unused events: `clear_notifications`, `update_points`
- Publish recorder events only to the Administrator
- Re-try the SocketIO client connection thrice (instead of infinitely) before stopping

gavindsouza committed:
chore: Remove unused realtime updates

These events were added for supporting listeners in desk. The listeners
have thus been removed and these are now unnecessary messages published
to anyone landing on Frappe pages or on Desk.

gavindsouza committed:
fix(recorder): Publish update only to Administrator

gavindsouza committed:
fix(socketio): Re-try thrice before trying to reconnect

Set reconnectionAttempts to 3. If the server doesn't want to connect
with the respective client (invalid origin, no cookie or sid transmitted)
or is gone down, socketio client would retry connections indefinitely.
This change limits retrying connection to just thrice every second.

gavindsouza committed:
refactor: SocketIO

- Check request data in middleware
- Authenticate each connection before allowing room access
- Allow site room access only to System Users, restrict Website User &
  Guests to their respective user rooms

Note: This doesn't check for roles / permissions

gavindsouza committed:
perf(socketio): get_site_name

The usages and number of conditions evaluated in the function called for
some sort of a cache. If the site name is evaluated once, store it in
the socket object!

gavindsouza committed:
chore: Drop duplicate event method

This particular definition was chosen since there was no corresponding
subscribe method with the same key generation logic

gavindsouza committed:
refactor(socketio)!: list_update

- Subscribe to list_update only for the list/report views that are
  opened
- Check if user has read permission for doctype to subscribe to list
  updates

gavindsouza committed:
refactor(socketio): docinfo_update

- Rename event from `update_docinfo_for_{}_{}` to docinfo_update
- Separate rooms for separate documents generated on requirement
- Check if user has access to doc before sharing docinfo

sagarvora commented on frappe/public/js/frappe/form/form.js:
```diff
@@ -1942,10 +1942,9 @@ frappe.ui.form.Form = class FrappeForm {
-		frappe.realtime.on(listener_name, ({ doc, key, action = "update" }) => {
+
+		frappe.socketio.docinfo_subscribe(doctype, docname);
+		frappe.realtime.on("docinfo_update", ({ doc, key, action = "update" }) => {
```
since this event is same for all forms now, this handler should only be "attached" once. needs redesign.

gavindsouza replied:
Fixed in 3a8fa6c

sagarvora replied:
fix won't work as expected. the vars `this`, `doctype` and `docname` will always refer to the current object if I'm not mistaken.

gavindsouza replied:
Yep, `this` was a mess 😄. Instead, I've now gone back to resetting and attaching a new callback. To ensure we aren't showing stale data, I'm popping the old document from locals too (desk makes a getdoc request then) [via 2b7e4554c4e0dc6efeffecb5e699e1a666388190]

sagarvora commented on socketio.js:
```diff
@@ -210,33 +239,38 @@ function get_typing_room(socket, doctype, docname) {
 	return get_site_name(socket) + ":all";
 }
 
+function get_list_room(socket, doctype) {
```
or it could be.. a "doctype" room.

gavindsouza replied:
Fixed in 3a8fa6c

gavindsouza committed:
refactor(socketio): Use same room for doc & info events

other changes
- Name list room as doctype room for more generic use
- avoid re-setting up listeners for generic events
- discard docinfo_subscribe event

gavindsouza committed:
Merge branch 'develop' into socket_io-refactor

ankush commented on frappe/public/js/frappe/socketio_client.js:
```diff
@@ -17,14 +20,17 @@ frappe.socketio = {
 		} else if (window.location.protocol == "file:") {
 			frappe.socketio.socket = io.connect(window.localStorage.server, {
 				withCredentials: true,
+				reconnectionAttempts: 3,
```
is connecting over file protocol even required?

gavindsouza replied:
![prattypratprat](https://user-images.githubusercontent.com/36654812/202119070-4a72f01b-7734-4f5a-a45d-182120dfe5df.png)

ankush replied:
lets remove this https://github.com/gavindsouza/frappe/commit/84f8661ca869b0f0ecddd02fa23f1e9507408c50

ankush committed:
chore: remove dead code

gavindsouza committed:
fix(desk): maintain realtime & cached data consistency

- Clear docinfo_update callbacks before setting one; ensure only
  one active callback at any given point.
- Remove document from local cache if list_update sent if not edited

gavindsouza commented:
Discussions broke (see Cypress) due to changes. If there are no major concerns or issues related to the changes, I'll fix discussions' realtime bits.

Update: Fixed by auto-joining all users to the respective site's room `website` by default. Teams developing website projects can / should make the call whenever the time comes on making this change.

gavindsouza committed:
feat: {site}:website room open to all users

- Subscribe to room and pass messages without auth
- Pass `room='website'` to publish_realtime to use
- Pass discussions' comms through particular site's website room

gavindsouza committed:
fix(socketio): Scoping & hoisting bugs

Due to the previous logic, cookie data seemed inconsistent causing ghost
sessions.

gavindsouza committed:
fix: Restrict socket data to respective users after commit

Fix conditions to bother only those who asked for the data:
- Clear permissions cache only for updated users' data
- Defer appropriate events until commit to avoid ghost events
- Remove event unused by desk (and other apps)

gavindsouza committed:
Merge branch 'develop' into socket_io-refactor

ankush commented on frappe/desk/doctype/notification_log/notification_log.py:
```diff
@@ -176,7 +176,7 @@ def mark_as_read(docname):
 @frappe.whitelist()
 def trigger_indicator_hide():
-	frappe.publish_realtime("indicator_hide", user=frappe.session.user)
+	frappe.publish_realtime("indicator_hide", user=frappe.session.user, after_commit=True)
```
```suggestion
	frappe.publish_realtime("indicator_hide", user=frappe.session.user)
```

This might not trigger a commit. Similar problem in other `after_commit=True`

ankush commented on frappe/public/js/frappe/model/model.js:
```diff
@@ -79,6 +79,10 @@ $.extend(frappe.model, {
 
 	init: function () {
 		// setup refresh if the document is updated somewhere else
+		if (frappe.socketio.is_document_listener_setup) {
```
We can remove this for now, `model.init` is only called once and rn only does doc_update listener.

ankush commented on frappe/public/js/frappe/socketio_client.js:
```diff
@@ -3,6 +3,9 @@ frappe.socketio = {
 	open_tasks: {},
 	open_docs: [],
 	emit_queue: [],
+	is_docinfo_listener_setup: false,
+	is_document_listener_setup: false,
```
```suggestion
```

ankush commented on frappe/realtime.py:
```diff
@@ -117,27 +117,46 @@ def can_subscribe_doc(doctype, docname):
 
+@frappe.whitelist(allow_guest=True)
+def can_subscribe_list(doctype):
+	if os.environ.get("CI"):
+		return True
+
```
```suggestion
```

gavindsouza committed:
fix(socketio): Revert irrelevant & unused changes

Co-authored-by: Ankush Menat <ankushmenat@gmail.com>

gavindsouza committed:
Merge branch 'develop' into socket_io-refactor

gavindsouza committed:
fix: Auto-add all users to website room

Keep website room open for all - currently only used for discussions
